rst_files/linear_algebra.rst

```diff
@@ -336,55 +314,29 @@ The span is a 2 dimensional plane passing through these two points and the origi
 
     α, β = 0.2, 0.1
 
-    # Axes
-    gs = 3
-    z = range(x_min, stop = x_max, length = gs)
-    x = zeros(gs)
-    y = zeros(gs)
-    plot(x, y, z, color = :black, linewidth=2, alpha=0.5, label = "", legend=false)
-    plot!(z, x, y, color = :black, linewidth=2, alpha=0.5, label = "")
-    plot!(y, z, x, color = :black, linewidth=2, alpha=0.5, label = "")
-
     # Fixed linear function, to generate a plane
-    f(x, y) = α .* x + β .* y
+    f(x, y) = α * x + β * y
 
     # Vector locations, by coordinate
     x_coords = [3, 3]
     y_coords = [4, -4]
-    z = f(x_coords, y_coords)
+    z_coords = f(x_coords, y_coords)
 
     # Lines to vectors
-    n = 2
-    x_vec = zeros(n, n)
-    y_vec = zeros(n, n)
-    z_vec = zeros(n, n)
-    labels = []
-
-    for i ∈ 1:n
-        x_vec[:, i] = [0; x_coords[i]]
-        y_vec[:, i] = [0; y_coords[i]]
-        z_vec[:, i] = [0; f(x_coords[i], y_coords[i])]
-        lab = string("a", i)
-        push!(labels, lab)
-    end
-
-    plot!(x_vec, y_vec, z_vec, color = [:blue :red], linewidth = 1.5,
-        alpha = 0.6, label = labels)
+    x_vec = reduce(hcat, ([0, x] for x in x_coords))
+    y_vec = reduce(hcat, ([0, y] for y in y_coords))
+    z_vec = reduce(hcat, ([0, z] for z in z_coords))
 
     # Draw the plane
     grid_size = 20
-    xr2 = range(x_min, stop = x_max, length = grid_size)
-    yr2 = range(y_min, stop = y_max, length = grid_size)
+    xr2 = range(x_min, x_max, length = grid_size)
+    yr2 = range(y_min, y_max, length = grid_size)
     z2 = zeros(grid_size, grid_size)
-    for i ∈ 1:grid_size
-        for j ∈ 1:grid_size
-            z2[j, i] = f(xr2[i], yr2[j])
-        end
+    for (position, value) in zip(eachindex(z2), (f(x, y) for x in xr2 for y in yr2))
```

Comment:
I have no idea what this code does.  Lets go back to some variation of the for loop, which is crystal clear.

Reply:
That's the Pythonic `[ f(x,y) for x in x_vals for y in y_vals ]`. It is the same as
```
reshape([ f(x,y) for x in x_vals for y in y_vals ], grid_size, grid_size)
```
I can go back to the nested for loops.

Reply:
```
  for j ∈ 1:grid_size
            z2[j, i] = f(xr2[i], yr2[j])
        end
```
or something like that just seems so easy to mentally parse here.  `reshape` is a tough concept for people

Reply:
That will just give a `Vector` and not the `Matrix`, hence the `reshape`.
The alternative is what I had of having an iterator with all the values in order and fill the `Matrix` one by one in the efficient order (relying on `eachindex(::Array)`).

Reply:
Check this out!
```
[ f(x,y) for x in x_vals, y in y_vals ]
```
Much better than a for loop in this case.

Reply:
(note the change of the `for` to a comma)?  I think this works?

Reply:
@jlperla This works, but it does a Cartesian grid instead of a pairwise thing. 

```
julia> [(a, b) for a in 1:5, b in 1:5]
5×5 Array{Tuple{Int64,Int64},2}:
 (1, 1)  (1, 2)  (1, 3)  (1, 4)  (1, 5)
 (2, 1)  (2, 2)  (2, 3)  (2, 4)  (2, 5)
 (3, 1)  (3, 2)  (3, 3)  (3, 4)  (3, 5)
 (4, 1)  (4, 2)  (4, 3)  (4, 4)  (4, 5)
 (5, 1)  (5, 2)  (5, 3)  (5, 4)  (5, 5)
```

Reply:
That is because `(a, b)` returns a tuple.  try it with `a + b` which returns an integer.

Reply:
Same story:

```
julia> [a+b for a in 1:5, b in 1:5]
5×5 Array{Int64,2}:
 2  3  4  5   6
 3  4  5  6   7
 4  5  6  7   8
 5  6  7  8   9
 6  7  8  9  10
```

It actually looks like a Cartesian grid is what we want in this case, unless I'm mistaken. But good to know for this pattern in the future.

Reply:
I am not sure what you mean by pairwise, but I guess it depends on what that `for` loop was doing.  I was responding on the fact that you couldn't get a matrix from a comprehension, which you can.

Regarldess, the pattern may be useful, as might taking the transpose of it.  This is a perfect example of where a few words on the algorithm explaining what we want and what the code needs to do could help.